connect/runtime/src/main/java/org/apache/kafka/connect/storage/ConnectorOffsetBackingStore.java

```diff
@@ -279,10 +284,61 @@ public Future<Void> set(Map<ByteBuffer, ByteBuffer> values, Callback<Void> callb
+                try (LoggingContext context = loggingContext()) {
+                    if (secondaryWriteError != null) {
+                        log.warn("Failed to write offsets with tombstone records to secondary backing store", secondaryWriteError);
+                        secondaryStoreTombstoneWriteError.compareAndSet(null, secondaryWriteError);
+                    } else {
+                        log.debug("Successfully flushed tombstone offsets to secondary backing store");
+                    }
+                }
+            });
+            try {
+                // For EOS, there is no timeout for offset commit and it is allowed to take as much time as needed for
+                // commits. We still need to wait because we want to fail the offset commit for cases when
+                // tombstone records fail to be written to the secondary store. Note that while commitTransaction
+                // already waits for all records to be sent and ack'ed, in this case we do need to add an explicit
+                // blocking call. In case of ALOS, we wait for the same duration as `offset.commit.timeout.ms`
+                // and throw that exception which would allow the offset commit to fail.
+                if (isEOSEnabled) {
+                    secondaryWriteFuture.get();
+                } else {
+                    secondaryWriteFuture.get(offsetFlushTimeoutMs, TimeUnit.MILLISECONDS);
+                }
+            } catch (InterruptedException e) {
+                log.warn("{} Flush of tombstone offsets to secondary store interrupted, cancelling", this);
+                secondaryStoreTombstoneWriteError.compareAndSet(null, e);
+            } catch (ExecutionException e) {
+                log.error("{} Flush of tombstone offsets to secondary store threw an unexpected exception: ", this, e);
+                secondaryStoreTombstoneWriteError.compareAndSet(null, e);
+            } catch (TimeoutException e) {
+                log.error("{} Timed out waiting to flush offsets with tombstones to secondary storage ", this);
+                secondaryStoreTombstoneWriteError.compareAndSet(null, e);
+            } catch (Exception e) {
+                log.error("{} Got Exception when trying to flush tombstone offsets to secondary storage", this);
+                secondaryStoreTombstoneWriteError.compareAndSet(null, e);
+            }
+            Throwable writeError = secondaryStoreTombstoneWriteError.get();
+            if (writeError != null) {
+                FutureCallback<Void> failedWriteCallback = new FutureCallback<>(callback);
+                failedWriteCallback.onCompletion(writeError, null);
+                return failedWriteCallback;
+            }
+        }
+
         return primaryStore.set(values, (primaryWriteError, ignored) -> {
-            if (secondaryStore != null) {
+            // Secondary store writes have already happened for tombstone records
```

Comment:
```suggestion
            // Secondary store writes have already happened if the offsets batch contains tombstone records
```
nit: the existing comment could be interpreted as - `only tombstone records in the offsets batch have been written to the secondary store`, which is not the case.